include/mbedtls/rsa.h

```diff
@@ -575,25 +409,41 @@ int mbedtls_rsa_gen_key( mbedtls_rsa_context *ctx,
 int mbedtls_rsa_check_pubkey( const mbedtls_rsa_context *ctx );
 
 /**
- * \brief          Check if a context contains an RSA private key
- *                 and perform basic sanity checks.
- *
- * \param ctx      RSA context to be checked
- *
- * \return         0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code.
- *                 On success, it is guaranteed that enough information is
- *                 present to perform RSA private and public key operations.
- *
- * \warning        This function is *not* obliged to perform an exhaustive
- *                 sanity check what would guarantee the internal parameters
- *                 to match and \c mbedtls_rsa_private and \c mbedtls_rsa_public
- *                 to be mutually inverse to each other.
- *                 The reason is that for minimal non-CRT implementations
- *                 using only N, D, E, for example, checking the validity
- *                 would be computationally expensive.
- *                 Users mistrusting their key material should use other
- *                 means for verification; see the documentation of
- *                 \c mbedtls_rsa_complete.
+ * \brief      Check if a context contains an RSA private key
+ *             and perform basic consistency checks.
+ *
+ * \param ctx  RSA context to be checked
+ *
+ * \return     0 if successful, or an \c MBEDTLS_ERR_RSA_XXX error code.
+ *
+ * \note       This function performs checks substantiating
```

Comment:
No, that's what `rsa_complete` guarantees (and cannot be bypassed), while this function is supposed to give confidence that the key is functionally sane.

Reply:
@mpg See https://github.com/ARMmbed/mbedtls/pull/1060#discussion_r143390209 I don't feel comfortable not providing any private key check which goes beyond "this won't crash". The suggestion, as implemented currently, is to have the unavoidable `rsa_complete` check that the key material is sufficient to run the RSA primitives without crash - this includes both public and private key cases - while `rsa_check_privkey` performs 'functional sanity' checks indicating that the parameters indeed make the public and private primitives inverses of each other. The perhaps convoluted formulation in the documentation was chosen because I want to include both the previous version of `rsa_check_privkey` which did a full consistency check of the parameters (and which is kept at the moment), but also a potential version of `rsa_check_privkey` for contexts having `N,D,E` only, where an example decryption-encryption would be more appropriate then trying to derive `P,Q` first and performing a full consistency check. 

I know the documentation has been blurry previously, giving us quite some freedom on what to do with `check_privkey`, but I also don't feel comfortable changing from a function with a previously very strong semantics (full consistency check) to one providing nothing but 'this won't crash' - and I also don't see the need, because `rsa_complete` is a natural candidate for the latter.

Reply:
Ok, thanks for clarifying. I agree with your point that we should not loosen the semantics of `check_privkey()` compared to what they are now.

In the current documentation, I don't really like the reference to `validate_params()` in the first note (the one in the warning is fine) because it looks like we're defining a public function in terms of an internal one, and similarly the end of the paragraph is more about how the function might be implemented. I think we should stick to describing the observable behaviour first, and implementation notes, if any, only later.

How about something like: "The consistency checks performed ensure not only that `rsa_private()` can be called successfully on this context, but also that the various parameters are consistent with each other, in the sense they make `rsa_public()` the inverse of `rsa_private()`." Then follow up with the warning, which IMO is clear as it is (plus Gilles' style improvements).

Reply:
Thanks @mpg, I took your suggestion with slight alterations to include empirical checks:
```
 * \note       The consistency checks performed by this function not only
 *             ensure that \c mbedtls_rsa_private can be called successfully
 *             on the given context, but that the various parameters are
 *             mutually consistent with high probability, in the sense that
 *             \c mbedtls_rsa_public and \c mbedtls_rsa_private are inverses.
```